File: backend/app/middleware/rate_limit.py
import time

from fastapi import HTTPException, status
from redis.asyncio import Redis

from app.config import get_settings

settings = get_settings()

_redis: Redis | None = None


async def get_redis() -> Redis:
    global _redis
    if _redis is None:
        _redis = Redis.from_url(settings.redis_url, decode_responses=True)
    return _redis


async def close_redis():
    global _redis
    if _redis is not None:
        await _redis.aclose()
        _redis = None


async def check_rate_limit(
    key: str,
    max_requests: int,
    window_seconds: int,
) -> tuple[bool, int]:
    """Sliding window rate limiter using Redis sorted sets."""
    redis = await get_redis()
    now = time.time()
    window_start = now - window_seconds

    pipe = redis.pipeline(transaction=True)
    pipe.zremrangebyscore(key, 0, window_start)
    pipe.zadd(key, {str(now): now})
    pipe.zcard(key)
    pipe.expire(key, window_seconds)
    results = await pipe.execute()

    request_count = results[2]
    remaining = max(0, max_requests - request_count)

    return request_count <= max_requests, remaining


async def rate_limit_user(user_id: str):
    """Rate limit per authenticated user."""
    key = f"rl:user:{user_id}"
    allowed, remaining = await check_rate_limit(
        key, settings.rate_limit_requests, settings.rate_limit_window_seconds
    )
    if not allowed:
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail="Rate limit exceeded. Please wait before making more requests.",
            headers={"Retry-After": str(settings.rate_limit_window_seconds)},
        )


async def rate_limit_login(identifier: str):
    """Rate limit login/register attempts per IP or email."""
    key = f"rl:login:{identifier}"
    allowed, remaining = await check_rate_limit(key, settings.rate_limit_login_attempts, 900)
    if not allowed:
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail="Too many attempts. Please try again later.",
            headers={"Retry-After": "900"},
        )


async def rate_limit_api_key_auth(ip: str):
    """Rate limit failed API key authentication attempts per IP.
    Prevents brute-force key guessing.
    """
    key = f"rl:apikey:{ip}"
    allowed, remaining = await check_rate_limit(key, 10, 60)  # 10 failures per minute
    if not allowed:
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail="Too many authentication failures.",
            headers={"Retry-After": "60"},
        )

File: backend-rust/src/services/rate_limit.rs
//! Redis-based sliding-window rate limiter.
//!
//! Uses a sorted set per key where each member is a unique request ID scored by
//! its Unix-microsecond timestamp.  An atomic `MULTI/EXEC` pipeline trims
//! expired entries, inserts the new one, counts the window, and sets the TTL —
//! all in a single round-trip.

use sha2::{Digest, Sha256};
use uuid::Uuid;

use crate::config::Config;

// ── Core rate-limit check ───────────────────────────────────────────────────

/// Check whether a request identified by `key` is within its rate limit.
///
/// Returns `(allowed, remaining)` where `remaining` is the number of requests
/// still available in the current window.
pub async fn check_rate_limit(
    redis: &redis::Client,
    key: &str,
    max_requests: i64,
    window_seconds: i64,
) -> Result<(bool, i64), RateLimitError> {
    let mut conn = redis
        .get_multiplexed_async_connection()
        .await
        .map_err(|e| RateLimitError::Redis(e.to_string()))?;

    let now_us = chrono::Utc::now().timestamp_micros();
    let window_us = window_seconds * 1_000_000;
    let min_score = now_us - window_us;

    // Unique member for this request
    let member = Uuid::new_v4().to_string();

    // Atomic pipeline: ZREMRANGEBYSCORE, ZADD, ZCARD, EXPIRE
    let result: ((), (), i64, ()) = redis::pipe()
        .atomic()
        .cmd("ZREMRANGEBYSCORE")
        .arg(key)
        .arg("-inf")
        .arg(min_score)
        .ignore()
        .cmd("ZADD")
        .arg(key)
        .arg(now_us)
        .arg(&member)
        .ignore()
        .cmd("ZCARD")
        .arg(key)
        .cmd("EXPIRE")
        .arg(key)
        .arg(window_seconds + 1) // slight buffer
        .ignore()
        .query_async(&mut conn)
        .await
        .map_err(|e| RateLimitError::Redis(e.to_string()))?;

    let count = result.2;
    let allowed = count <= max_requests;
    let remaining = (max_requests - count).max(0);

    Ok((allowed, remaining))
}

// ── Convenience helpers ─────────────────────────────────────────────────────

/// Check the general per-user rate limit; returns `Err(RateLimitError::Exceeded)`
/// if the user has exhausted their quota.
pub async fn rate_limit_user(
    redis: &redis::Client,
    user_id: &str,
    config: &Config,
) -> Result<(), RateLimitError> {
    let key = format!("rl:user:{user_id}");
    let (allowed, remaining) = check_rate_limit(
        redis,
        &key,
        config.rate_limit_requests as i64,
        config.rate_limit_window_seconds as i64,
    )
    .await?;

    if !allowed {
        tracing::warn!(
            user_id = user_id,
            remaining = remaining,
            "user rate limit exceeded"
        );
        return Err(RateLimitError::Exceeded {
            retry_after_seconds: config.rate_limit_window_seconds,
        });
    }

    Ok(())
}

/// Check the login-specific rate limit (5 attempts per 900 s).
pub async fn rate_limit_login(
    redis: &redis::Client,
    identifier: &str,
) -> Result<(), RateLimitError> {
    let key = format!("rl:login:{identifier}");
    let (allowed, _remaining) = check_rate_limit(redis, &key, 5, 900).await?;

    if !allowed {
        tracing::warn!(
            identifier_hash = %_hash_short(identifier),
            "login rate limit exceeded"
        );
        return Err(RateLimitError::Exceeded {
            retry_after_seconds: 900,
        });
    }

    Ok(())
}

/// Check the API-key authentication rate limit (10 per 60 s, keyed by IP).
pub async fn rate_limit_api_key_auth(
    redis: &redis::Client,
    ip: &str,
) -> Result<(), RateLimitError> {
    let key = format!("rl:apikey:{ip}");
    let (allowed, _remaining) = check_rate_limit(redis, &key, 10, 60).await?;

    if !allowed {
        tracing::warn!(ip = ip, "api-key auth rate limit exceeded");
        return Err(RateLimitError::Exceeded {
            retry_after_seconds: 60,
        });
    }

    Ok(())
}

// ── Error type ──────────────────────────────────────────────────────────────

/// Errors produced by the rate-limiting layer.
#[derive(Debug, thiserror::Error)]
pub enum RateLimitError {
    #[error("rate limit exceeded — retry after {retry_after_seconds}s")]
    Exceeded { retry_after_seconds: u64 },

    #[error("redis error: {0}")]
    Redis(String),
}

// ── Helpers ─────────────────────────────────────────────────────────────────

/// Short SHA-256 prefix for logging identifiers without leaking PII.
fn _hash_short(value: &str) -> String {
    let digest = Sha256::digest(value.as_bytes());
    digest
        .iter()
        .take(4)
        .fold(String::with_capacity(8), |mut s, b| {
            use std::fmt::Write;
            write!(s, "{b:02x}").unwrap();
            s
        })
}
